src/individual/CalcBaseDamage.c

```diff
@@ -225,836 +310,1421 @@ int CalcBaseDamage(void *bw, struct BattleStruct *sp, int moveno, u32 side_cond,
-            {
-                damage = damage * 2 / 3;
+        if (defender == sp->rawSpeedNonRNGClientOrder[i]) {
+            // handle Heatproof
+            // https://www.smogon.com/forums/threads/scarlet-violet-battle-mechanics-research.3709545/post-9800504
+            if ((movetype == TYPE_FIRE)
+            && (MoldBreakerAbilityCheck(sp, attacker, defender, ABILITY_HEATPROOF) == TRUE)) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__0_5);
+                continue;
             }
-            else
-            {
-                damage /= 2;
+
+            // handle Thick Fat
+            if ((MoldBreakerAbilityCheck(sp, attacker, defender, ABILITY_THICK_FAT) == TRUE)
+            && ((movetype == TYPE_FIRE) || (movetype == TYPE_ICE))) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__0_5);
             }
-        }
-    }
-    else// if (movesplit == SPLIT_SPECIAL) // same as above, handle special moves
-    {
-        // handle light screen
-        if (((side_cond & SIDE_STATUS_LIGHT_SCREEN) != 0)
-         && (critical == 1)
-         && (sp->moveTbl[moveno].effect != MOVE_EFFECT_REMOVE_SCREENS)
-         && (AttackingMon.ability != ABILITY_INFILTRATOR))
-        {
-            if ((battle_type & BATTLE_TYPE_DOUBLE) && (CheckNumMonsHit(bw, sp, 1, defender) == 2))
-            {
-                damage = damage * 2 / 3;
+
+            // handle Water Bubble
+            if ((DefendingMon.ability == ABILITY_WATER_BUBBLE) && (movetype == TYPE_FIRE)) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__0_5);
             }
-            else
-            {
-                damage /= 2;
+
+            // handle Purifying Salt
+            // https://www.smogon.com/forums/threads/scarlet-violet-battle-mechanics-research.3709545/post-9469856
+            if ((DefendingMon.ability == ABILITY_PURIFYING_SALT) && (movetype == TYPE_GHOST)) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__0_5);
             }
         }
     }
 
-    if ((battle_type & BATTLE_TYPE_DOUBLE) &&
-        (sp->moveTbl[moveno].target == 0x4) &&
-        (CheckNumMonsHit(bw, sp, 1, defender) == 2))
-    {
-        damage = damage * 3 / 4;
-    }
+    // Items
+    for (i = 0; i < maxBattlers; i++) {
+        if (attacker == sp->rawSpeedNonRNGClientOrder[i]) {
+            // handle Choice Band
+            if ((AttackingMon.item_held_effect == HOLD_EFFECT_CHOICE_ATK) && (movesplit == SPLIT_PHYSICAL)) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__1_5);
+            }
+            
+            // handle Choice Specs
+            if ((AttackingMon.item_held_effect == HOLD_EFFECT_CHOICE_SPATK) && (movesplit == SPLIT_SPECIAL)) {
+                attackModifier = QMul_RoundUp(attackModifier, UQ412__1_5);
+            }
+    
+            // TODO: it’s not a Ditto/Smeargle/Mew Transformed into the species <- how to do this?
```

Comment:
should be handled by just checking the species?

Reply:
I think vanilla Gen 4 doesn't store the pre-transformed species?